src/dmd/lexer.d

```diff
@@ -202,6 +203,104 @@ unittest
     }
 }
 
+
+/**************
+ * Pool allocator for Token objects, used by the Lexer.
+ * Tokens are never freed so Token pointers will never change
+ */
+private struct TokenPool
+{
+    private Array!Token tokens;
+
+    public ref Token getByIndex(size_t index) nothrow pure
+    {
+        if (index >= tokens.length)
+        {
+            tokens.setDim(index + 1);
+        }
+        return tokens[index];
+    }
+}
+
+
+/***************
+ * Represents a range of tokens lazily scanned from a lexer
+ * should be created with the Lexer makeRangeFromHere method
+ */
+struct TokenRange
+{
+private:
+    Lexer lex; //where are these tokens coming from?
```

Comment:
BTW, is there a need to make this private since there's a getter available?

Reply:
I don't think you should be able to change which lexer the range is attached to.  It's basically logically const, but the methods I call might affect the lexer's "read ahead buffer" so it can't be actually const.

Reply:
Fair enough.